**Inclass material for Week 4: SQL Query using `pandas`**

This notebook was made based on main materials `4_SQL_Query.ipynb`

Version: Theia - September 2022

---
**START OF DAY 1**

# SQL Query using `pandas`

**Training Objectives**

- Querying from SQL Databases
- SQL Joins
- SQL Conditional Statements
- Flavors and Common Operators
- End to end data analysis

# Working with SQL Databases

## Database Schema

Database `chinook.db` terdiri dari 11 tabel yang berisi sampel data dari sebuah toko yang menjual media digital:

- Data terkait media seperti pada tabel `tracks`, `albums`, `artists`, `genres`, `media_types`, `playlists`, `playlist_track` merupakan data asli dari library Apple iTunes.
- Informasi mengenai `customers` dan `employees` dibuat menggunakan nama dan alamat fiktif yang dapat ditemukan di Google maps, dan data lain yang diformat dengan baik (telepon, fax, email, dll).
- Informasi penjualan `invoices` dan `invoice_items` dihasilkan secara otomatis menggunakan data acak untuk periode tahun 2009-2013.

Skema berikut sering disebut sebagai **Entity Relationship Diagram (ERD)**, menunjukkan:
1. Entitas (Tabel)
2. Atribut (Kolom beserta tipe datanya)
3. Kardinalitas (Hubungan antar tabel)

![](assets/chinookschema2.png)

Gambar di atas adalah ERD untuk database chinook:

- Satu database terdiri dari beberapa tabel
- Setiap tabel terdiri dari beberapa atribut/kolom
- Terdapat kolom unik yang membedakan satu tabel dengan lainnya, disebut sebagai **Primary Key (PK)**. PK dapat digunakan sebagai kolom penghubung antar tabel saat dilakukan proses join (penggabungan) tabel.

## Database Connection

Sedikit berbeda dengan `pandas` yang semua operasinya dilakukan di komputer lokal Anda. Ketika Anda bekerja dengan SQL, kemungkinan besar Anda memiliki database relasional yang disimpan **terpusat** pada server dan dapat **diakses oleh beberapa client**. Berikut adalah ilustrasinya:

![](assets/clientserver.png)

Untuk menghubungkan client machine dengan server, kita perlu membuat sebuah connection melalui sebuah DBAPI. Pada Python, DBAPI ekuivalen dengan sebuah package yang memungkinkan seorang data analyst dapat membuat objek **database connection**, salah satunya adalah `sqlite3`. Berikut adalah ilustrasi dari DBAPI:

![](assets/dbapi.png)

**[OPSIONAL]** Berikut adalah contoh package lain pada Python untuk mengkoneksikan dengan database:

<br>

<details>
    <summary>✨ Connecting to MySQL</summary>
    
```python
import pymysql
  
conn = pymysql.connect(
    host = HOST_NAME,
    port = PORT_NUMBER,
    user = USER_NAME,
    password = PASSWORD,
    db = DATABASE_NAME)
```
</details>

<br>

<details>
    <summary>✨ Connecting to Oracle</summary>
    
```python
import cx_Oracle
  
# data source name from tnsnames.ora file
dsn_tns = cx_Oracle.makedsn(
    HOST_NAME,
    PORT_NUMBER
    service_name = SERVICE_NAME)

# connection
conn = cx_Oracle.connect(
    user = USER_NAME,
    password = PASSWORD,
    dsn = dsn_tns)
```
</details>

<br>

<details>
    <summary>✨ Connecting to PostgreSQL</summary>
    
```python
import psycopg2

conn = psycopg2.connect(
    host = HOST_NAME,
    port = PORT_NUMBER,
    user = USER_NAME,
    password = PASSWORD,
    database = DATABASE_NAME)
```
</details>

<br>

<details>
    <summary>✨ Connecting to Microsoft SQL Server</summary>
    
```python
import pyodbc 
conn = pyodbc.connect(
    'Driver={ODBC Driver 17 for SQL Server};'
    'Server=host;'
    'PORT=1433;'
    'UID=user;'
    'PWD=password;'
    'Database=database;')
```
</details>

Referensi connect DB dengan MySQL ke Python: https://youtu.be/h_gQ470uArc

Sebagai tahap awal, mari kita coba bagaimana mengkoneksikan Jupyter Notebook dengan database SQLite (menggunakan package `sqlite3`) yang disebut sebagai **connection**:

In [38]:
import sqlite3 # default package: Database API (DBAPI) / package
import pandas as pd

In [39]:
# membuat objek connection
conn = sqlite3.connect("data_input/chinook.db")
conn

## `SELECT` Statements

`SELECT` digunakan untuk memilih **kolom** dari sebuah **tabel**.

Syntax `SELECT`:

```sql
SELECT <NAMA_KOLOM>
FROM <NAMA_TABLE>
```

Melalui `pandas` kita dapat menggunakan method method `pd.read_sql_query(sql, con)`. Parameter:

- `sql`: SQL query dalam bentuk string
- `con`: object connection

Misal kita hanya ingin mengambil kolom `AlbumId` dan `Title` dari tabel `albums`:

In [40]:
sql="SELECT * FROM albums"
albums = pd.read_sql_query(sql, conn)
albums

AlbumId                                              Title  ArtistId
0          1              For Those About To Rock We Salute You         1
1          2                                  Balls to the Wall         2
2          3                                  Restless and Wild         2
3          4                                  Let There Be Rock         1
4          5                                           Big Ones         3
..       ...                                                ...       ...
342      343                             Respighi:Pines of Rome       226
343      344  Schubert: The Late String Quartets & String Qu...       272
344      345                                Monteverdi: L'Orfeo       273
345      346                              Mozart: Chamber Music       274
346      347  Koyaanisqatsi (Soundtrack from the Motion Pict...       275

[347 rows x 3 columns]

Gunakan command `SELECT *` untuk mengambil **semua kolom** dari tabel `albums`:

In [41]:
sql="SELECT * FROM tracks"
tracks = pd.read_sql_query(sql, conn)
tracks.info()
# tracks['MediaTypeId'].unique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3508 entries, 0 to 3507
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   TrackId       3508 non-null   int64  
 1   Name          3508 non-null   object 
 2   AlbumId       3508 non-null   int64  
 3   MediaTypeId   3508 non-null   int64  
 4   GenreId       3508 non-null   int64  
 5   Composer      2530 non-null   object 
 6   Milliseconds  3508 non-null   int64  
 7   Bytes         3508 non-null   int64  
 8   UnitPrice     3508 non-null   float64
dtypes: float64(1), int64(6), object(2)
memory usage: 246.8+ KB


💡 **Note**: Berbeda dari Python, perintah yang ada di SQL bersifat case **insensitive**. Query di atas dapat juga dituliskan sebagai `select * from albums`. Namun, demi kemudahan pembacaan, statement biasa dituliskan dengan huruf besar (dalam hal ini `SELECT` dan `FROM`).

### ❓ Knowledge Check

Kali ini, coba ambil semua kolom pada tabel `artists` kemudian simpan ke dalam sebuah object dataframe bernama `artist`.

**Pertanyaan:** Berapa baris yang ada pada dataframe `artist`?

<br>

<details>
    <summary><i>👉 Klik di sini untuk hint</i></summary>
    
```python
___ = pd.read_sql_query("SELECT ___ FROM ___", conn)
```
</details>

In [42]:
# your code here
sql="SELECT * FROM artists"
artist = pd.read_sql_query(sql, conn)
artist

ArtistId                                               Name
0           1                                              AC/DC
1           2                                             Accept
2           3                                          Aerosmith
3           4                                  Alanis Morissette
4           5                                    Alice In Chains
..        ...                                                ...
270       271   Mela Tenenbaum, Pro Musica Prague & Richard Kapp
271       272                             Emerson String Quartet
272       273  C. Monteverdi, Nigel Rogers - Chiaroscuro; Lon...
273       274                                      Nash Ensemble
274       275                              Philip Glass Ensemble

[275 rows x 2 columns]

In [43]:
sql="SELECT * FROM invoices"
invoices = pd.read_sql_query(sql, conn, parse_dates='InvoiceDate')
invoices

InvoiceId  CustomerId InvoiceDate  \
0            1           2  2009-01-01   
1            2           4  2009-01-02   
2            3           8  2009-01-03   
3            4          14  2009-01-06   
4            5          23  2009-01-11   
..         ...         ...         ...   
407        408          25  2013-12-05   
408        409          29  2013-12-06   
409        410          35  2013-12-09   
410        411          44  2013-12-14   
411        412          58  2013-12-22   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart         None   
1                            Ullevålsveien 14        Oslo         None   
2                             Grétrystraat 63    Brussels         None   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto         None   
410                           Porthaninkatu 9    Helsinki         None   
411                       12,Community Centre       Delhi         None   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal              None   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

Berikut adalah parameter yang sering digunakan dalam method `pd.read_sql_query()`:

- `sql`: SQL query dalam bentuk string
- `con`: object connection
- `index_col`: nama atau index kolom yang ingin dijadikan index (seperti pada `pd.read_csv()`)
- `parse_dates`: nama kolom yang ingin dikonversi menjadi tipe data `datetime64` (seperti pada `pd.read_csv()`)

> Kunjungi [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql_query.html) untuk detail lebih lanjut.

### LIMIT

Statement `LIMIT` digunakan untuk mengambil beberapa baris pertama pada data.

Kita ingin ambil **5 baris pertama** dari tabel `artists` lalu set kolom `ArtistId` menjadi index.

In [44]:
# your code here
sql="SELECT * FROM tracks "\
    "INNER JOIN albums ON albums.AlbumId = tracks.AlbumId "\
    "INNER JOIN artists ON artists.ArtistId = albums.ArtistId"
tracks_artist = pd.read_sql_query(sql, conn)
tracks_artist

TrackId                                               Name  AlbumId  \
0           1            For Those About To Rock (We Salute You)        1   
1           6                              Put The Finger On You        1   
2           7                                    Let's Get It Up        1   
3           8                                   Inject The Venom        1   
4           9                                         Snowballed        1   
...       ...                                                ...      ...   
3503     3499  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...      343   
3504     3500  String Quartet No. 12 in C Minor, D. 703 "Quar...      344   
3505     3501               L'orfeo, Act 3, Sinfonia (Orchestra)      345   
3506     3502  Quintet for Horn, Violin, 2 Violas, and Cello ...      346   
3507     3503                                      Koyaanisqatsi      347   

      MediaTypeId  GenreId                                   Composer  \
0               1        1  Angus Young, Malcolm Young, Brian Johnson   
1               1        1  Angus Young, Malcolm Young, Brian Johnson   
2               1        1  Angus Young, Malcolm Young, Brian Johnson   
3               1        1  Angus Young, Malcolm Young, Brian Johnson   
4               1        1  Angus Young, Malcolm Young, Brian Johnson   
...           ...      ...                                        ...   
3503            2       24                                       None   
3504            2       24                             Franz Schubert   
3505            2       24                         Claudio Monteverdi   
3506            2       24                    Wolfgang Amadeus Mozart   
3507            2       10                               Philip Glass   

      Milliseconds     Bytes  UnitPrice  AlbumId  \
0           343719  11170334       0.99        1   
1           205662   6713451       0.99        1   
2           233926   7636561       0.99        1   
3           210834   6852860       0.99        1   
4           203102   6599424       0.99        1   
...            ...       ...        ...      ...   
3503        286741   4718950       0.99      343   
3504        139200   2283131       0.99      344   
3505         66639   1189062       0.99      345   
3506        221331   3665114       0.99      346   
3507        206005   3305164       0.99      347   

                                                  Title  ArtistId  ArtistId  \
0                 For Those About To Rock We Salute You         1         1   
1                 For Those About To Rock We Salute You         1         1   
2                 For Those About To Rock We Salute You         1         1   
3                 For Those About To Rock We Salute You         1         1   
4                 For Those About To Rock We Salute You         1         1   
...                                                 ...       ...       ...   
3503                             Respighi:Pines of Rome       226       226   
3504  Schubert: The Late String Quartets & String Qu...       272       272   
3505                                Monteverdi: L'Orfeo       273       273   
3506                              Mozart: Chamber Music       274       274   
3507  Koyaanisqatsi (Soundtrack from the Motion Pict...       275       275   

                                                   Name  
0                                                 AC/DC  
1                                                 AC/DC  
2                                                 AC/DC  
3                                                 AC/DC  
4                                                 AC/DC  
...                                                 ...  
3503                                     Eugene Ormandy  
3504                             Emerson String Quartet  
3505  C. Monteverdi, Nigel Rogers - Chiaroscuro; Lon...  
3506                                      Nash Ens

In [45]:

sql="SELECT * FROM artists limit 5"
artist_limit = pd.read_sql_query(sql, conn)
artist_limit

ArtistId               Name
0         1              AC/DC
1         2             Accept
2         3          Aerosmith
3         4  Alanis Morissette
4         5    Alice In Chains

## SQL Joins

Statement `JOIN` digunakan untuk **menggabungkan dua tabel** menjadi satu tabel melalui **kolom penghubung** yang sama. Operasi `JOIN` dapat dilakukan lebih dari satu kali dalam satu SQL query. Berikut adalah diagram ilustrasi beberapa macam operasi SQL `JOIN`:

![](assets/sqljoins.png)
Credit: Data & Object Factory, LLC

- `LEFT JOIN` paling sering digunakan pada sebagian besar skenario bisnis. `LEFT JOIN` mengembalikan semua baris pada tabel kiri **terlepas** dari apakah ada baris yang cocok pada tabel kanan.
- `INNER JOIN` adalah tipe join yang sangat intuitif dan mudah dipahami. Query ini mengembalikan semua baris di tabel kiri yang cocok dengan tabel kanan.

**Note**:

- `RIGHT JOIN` hampir tidak pernah digunakan karena ekuivalen dengan `LEFT JOIN` hanya mengganti peletakkan tabel kiri dan kanannya saja.
- `FULL OUTER JOIN` sangat jarang digunakan. Selain alasan komputasi, tipe join ini mengembalikan semua baris dari kedua tabel terlepas dari apakah ada kecocokan atau tidak, sehingga menghasilkan DataFrame dengan banyak missing value.

> Ilustrasi terkait ke-empat tipe join dapat dilihat pada [Visual JOIN](https://joins.spathon.com/).

Syntax `JOIN`:

```sql
SELECT <COLUMNS>
FROM <LEFT_TABLE>
[LEFT|INNER] JOIN <RIGHT_TABLE>
ON <LEFT_TABLE>.key = <RIGHT_TABLE>.key
```

**Kasus:** Kita ingin mendapatkan DataFrame yang menampilkan semua baris dengan kolom `AlbumId`, `Title`, dan `Name`. 

Perhatikan bahwa kolom `Name` terdapat pada tabel `artists`, sedangkan kolom `AlbumId` dan `Title` terdapat pada tabel `albums`. Berikut skema pada database `chinook.db`:

1. Tabel `albums`: 
    - `AlbumId`
    - `Title`
    - `ArtistId`


2. Tabel `artists`:
    - `ArtistId`
    - `Name` 

In [46]:
albums = pd.read_sql_query("SELECT AlbumId, Title, Name \
                            FROM albums \
                            LEFT JOIN artists \
                            ON artists.ArtistId = albums.ArtistId", conn)
albums.head()

AlbumId                                  Title       Name
0        1  For Those About To Rock We Salute You      AC/DC
1        2                      Balls to the Wall     Accept
2        3                      Restless and Wild     Accept
3        4                      Let There Be Rock      AC/DC
4        5                               Big Ones  Aerosmith

Perhatikan SQL query di atas:
- Tabel `albums` menjadi **tabel kiri**, sedangkan tabel `artists` menjadi **tabel kanan**.
- Kedua tabel digabungkan menggunakan operasi `LEFT JOIN` melalui key `ArtistId`.
- Penggunaan karakter backslash (`\`) ditujukan agar ke-empat baris SQL query dibaca sebagai satu kesatuan baris. Namun penggunaan backslash kurang efisien dan memiliki resiko error yang lebih tinggi. **Sebagai alternatif, Anda dapat mengapit SQL query dengan petik tiga kali `"""` atau `'''`**

In [47]:
# re-create tabel di atas menggunakan petik tiga kali 
albums = pd.read_sql_query('''SELECT AlbumId, Title, Name 
                            FROM albums 
                            LEFT JOIN artists 
                            ON artists.ArtistId = albums.ArtistId''', conn)
albums.head()

AlbumId                                  Title       Name
0        1  For Those About To Rock We Salute You      AC/DC
1        2                      Balls to the Wall     Accept
2        3                      Restless and Wild     Accept
3        4                      Let There Be Rock      AC/DC
4        5                               Big Ones  Aerosmith

Terkadang penamaan kolom ataupun tabel pada database cukup panjang dan redundan untuk diketik, ataupun belum cukup informatif.

Penggunaan statement `AS` dapat digunakan untuk melakukan **aliasing / rename** nama tabel maupun nama kolom.

In [48]:
# re-create tabel di atas menggunakan aliasing
albums = pd.read_sql_query("SELECT AlbumId, Title, Name \
                            FROM albums ab \
                            LEFT JOIN artists \
                            ON artists.ArtistId = ab.ArtistId", conn)
albums.head()

AlbumId                                  Title       Name
0        1  For Those About To Rock We Salute You      AC/DC
1        2                      Balls to the Wall     Accept
2        3                      Restless and Wild     Accept
3        4                      Let There Be Rock      AC/DC
4        5                               Big Ones  Aerosmith

> ⚠️ Hati-hati! Saat melakukan `LEFT JOIN`, peletakkan tabel kiri dan kanan sangat berpengaruh terhadap hasil penggabungan.

Sebagai perbanding, sekarang coba kita tukar tabel `artists` menjadi tabel kiri, sedangkan `albums` menjadi tabel kanan.

Object `x` di bawah menampilkan semua data `artists` walaupun artist tersebut tidak ada di tabel `albums`.

In [49]:
x = pd.read_sql_query(
    """
    SELECT AlbumId, Title, Name
    FROM artists
    LEFT JOIN albums
    ON artists.ArtistId = albums.ArtistId
    """, conn)

# cek baris dengan missing value
x[x.isna().any(axis=1)]
# x[x.isna().any(axis=1)]

AlbumId Title                                               Name
50       NaN  None                         Milton Nascimento & Bebeto
51       NaN  None                                            Azymuth
55       NaN  None                                      João Gilberto
56       NaN  None                                     Bebel Gilberto
57       NaN  None                                      Jorge Vercilo
..       ...   ...                                                ...
327      NaN  None                 DJ Dolores & Orchestra Santa Massa
328      NaN  None                                          Seu Jorge
329      NaN  None                               Sabotage E Instituto
330      NaN  None                                    Stereo Maracana
377      NaN  None  Academy of St. Martin in the Fields, Sir Nevil...

[71 rows x 3 columns]

### ❓ Knowledge Check

Buatlah sebuah object dataframe bernama `tracks` yang menampilkan:

- Semua kolom pada tabel `tracks`
- Kolom `Title` pada tabel `albums`
- Kolom `Name` pada tabel `artists`
- Kolom `Name` pada tabel `genres`

Lalu set kolom `TrackId` menjadi index pada dataframe tersebut. Apabila sudah benar, seharusnya dataframe Anda memiliki 3503 baris dan 11 kolom.

<br>

<details>
    <summary><i>👉 Klik di sini untuk hint</i></summary>

> **Hint 1**: Pada kasus ini, Anda wajib menggunakan syntax `<nama_tabel>.<nama_kolom>` pada statement `SELECT`. Hal ini karena terdapat nama kolom yang sama dari tabel yang berbeda. Jangan lupa untuk menggunakan aliasing `AS` agar nama kolom akhir tidak ambigu.
    
> **Hint 2**: Anda dapat menggunakan `SELECT tracks.*` untuk mengambil semua kolom pada tabel `tracks`

> **Hint 3**: Anda dapat melanjutkan hasil operasi `LEFT JOIN` dengan `LEFT JOIN` berikutnya. Sebagai referensi, berikut adalah struktur kodenya:
    
```python
___ = pd.read_sql_query(
    """
    SELECT ___
    FROM ___
    LEFT JOIN ___
    ON ___
    LEFT JOIN ___
    ON ___
    LEFT JOIN ___
    ON ___
    """, conn, index_col=___)
```
</details>

In [50]:
# your code here
tracks = pd.read_sql_query(
    """
    SELECT tracks.*, albums.Title, artists.Name as ArtistName, genres.Name as genreName
    FROM tracks
    LEFT JOIN albums ON albums.albumid = tracks.albumid
    LEFT JOIN artists ON albums.artistid = artists.artistid
    LEFT JOIN genres ON genres.genreid = tracks.genreid
    """, conn, index_col='TrackId')
tracks

Name  AlbumId  MediaTypeId  \
TrackId                                                                  
1        For Those About To Rock (We Salute You)        1            1   
2                              Balls to the Wall        2            2   
3                                Fast As a Shark        3            2   
4                              Restless and Wild        3            2   
5                           Princess of the Dawn        3            2   
...                                          ...      ...          ...   
3504                                    junkdata        1            1   
3505                                    junkdata        1            1   
3506                                    junkdata        1            1   
3507                                    junkdata        1            1   
3508                                    junkdata        1            1   

         GenreId                                           Composer  \
TrackId                                                               
1              1          Angus Young, Malcolm Young, Brian Johnson   
2              1                                               None   
3              1  F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...   
4              1  F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...   
5              1                         Deaffy & R.A. Smith-Diesel   
...          ...                                                ...   
3504           1                                           junkdata   
3505           1                                           junkdata   
3506           1                                           junkdata   
3507           1                                           junkdata   
3508           1                                           junkdata   

         Milliseconds     Bytes  UnitPrice  \
TrackId                                      
1              343719  11170334       0.99   
2              342562   5510424       0.99   
3              230619   3990994       0.99   
4              252051   4331779       0.99   
5              375418   6290521       0.99   
...               ...       ...        ...   
3504                1         1       5.00   
3505                1         1       5.00   
3506                1         1       5.00   
3507                1         1       5.00   
3508                1         1       5.00   

                                         Title ArtistName genreName  
TrackId                                                              
1        For Those About To Rock We Salute You      AC/DC      Rock  
2                            Balls to the Wall     Accept      Rock  
3                            Restless and Wild     Accept      Rock  
4                            Restless and Wild     Accept      Rock  
5                            Restless and Wild     Accept      Rock  
...                                        ...        ...       ...  
3504     For Those About To Rock We Salute You      AC/DC      Rock  
3505     For Those About To Rock We Salute You      AC/DC      Rock  
3506     For Those About To Rock We Salute You      AC/DC      Rock  
3507     For Those About To Rock We Salute You      AC/DC      Rock  
3508     For Those About To Rock We Salute You      AC/DC      Rock  

[3508 rows x 11 columns]

Gunakan object dataframe `tracks` yang telah Anda peroleh di atas dalam menjawab pertanyaan berikut:

1. Pilihlah semua nama genre yang terdapat pada 5 baris terakhir pada dataframe `tracks`. Jawaban bisa lebih dari satu.
    - [ ] Latin
    - [X] Classical
    - [X] Soundtrack
    - [ ] Pop

In [51]:
tracks['genreName'].tail(5)

TrackId
3504    Rock
3505    Rock
3506    Rock
3507    Rock
3508    Rock
Name: genreName, dtype: object

2. Di antara genre berikut, manakah yang termasuk dalam top 3 genre yang paling banyak muncul pada dataframe `tracks`? Buatlah sebuah tabel frekuensi dengan metode yang telah Anda pelajari!

    - [X] Latin
    - [ ] Classical
    - [ ] Soundtrack
    - [ ] Pop

<br>

<details>
    <summary><i>👉 Klik di sini untuk hint</i></summary>

Beberapa pilihan method yang dapat Anda gunakan:
    
- `.value_counts()`
- `pd.crosstab(index=___, columns=___)`
- `.pivot_table(index=___, aggfunc='count')`
- `.groupby().count()`
</details>

In [52]:
# your code here
tracks_top_3 = tracks.groupby('genreName').count()['Name']
# tracks_top_3.sort_values(ascending=False).head(3).plot.bar()

# tracks['genreName'].value_counts().head(3).plot.bar()
tracks

Name  AlbumId  MediaTypeId  \
TrackId                                                                  
1        For Those About To Rock (We Salute You)        1            1   
2                              Balls to the Wall        2            2   
3                                Fast As a Shark        3            2   
4                              Restless and Wild        3            2   
5                           Princess of the Dawn        3            2   
...                                          ...      ...          ...   
3504                                    junkdata        1            1   
3505                                    junkdata        1            1   
3506                                    junkdata        1            1   
3507                                    junkdata        1            1   
3508                                    junkdata        1            1   

         GenreId                                           Composer  \
TrackId                                                               
1              1          Angus Young, Malcolm Young, Brian Johnson   
2              1                                               None   
3              1  F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...   
4              1  F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...   
5              1                         Deaffy & R.A. Smith-Diesel   
...          ...                                                ...   
3504           1                                           junkdata   
3505           1                                           junkdata   
3506           1                                           junkdata   
3507           1                                           junkdata   
3508           1                                           junkdata   

         Milliseconds     Bytes  UnitPrice  \
TrackId                                      
1              343719  11170334       0.99   
2              342562   5510424       0.99   
3              230619   3990994       0.99   
4              252051   4331779       0.99   
5              375418   6290521       0.99   
...               ...       ...        ...   
3504                1         1       5.00   
3505                1         1       5.00   
3506                1         1       5.00   
3507                1         1       5.00   
3508                1         1       5.00   

                                         Title ArtistName genreName  
TrackId                                                              
1        For Those About To Rock We Salute You      AC/DC      Rock  
2                            Balls to the Wall     Accept      Rock  
3                            Restless and Wild     Accept      Rock  
4                            Restless and Wild     Accept      Rock  
5                            Restless and Wild     Accept      Rock  
...                                        ...        ...       ...  
3504     For Those About To Rock We Salute You      AC/DC      Rock  
3505     For Those About To Rock We Salute You      AC/DC      Rock  
3506     For Those About To Rock We Salute You      AC/DC      Rock  
3507     For Those About To Rock We Salute You      AC/DC      Rock  
3508     For Those About To Rock We Salute You      AC/DC      Rock  

[3508 rows x 11 columns]

3. Gunakan `groupby()` pada kolom `ArtistName` untuk menghitung `mean()` dari `UnitPrice`. Anda akan melihat bahwa mayoritas artists memiliki rata-rata `UnitPrice` di angka 0.99, namun ada juga beberapa artists yang rata-ratanya 1.99. Dari nama artists berikut, pilihlah yang memiliki rata-rata `UnitPrice` sebesar 0.99:

    - [ ] The Office
    - [ ] Aquaman
    - [X] Pearl Jam
    - [ ] Lost

In [53]:
# your code here
artist_mean = tracks.groupby(by='ArtistName').agg({'UnitPrice':'mean'})
artist_name = ['The Office','Aquaman','Pearl Jam','Lost'] 

for check in artist_name:
    chk_value = artist_mean.loc[check].values[0].round(2)
#     if chk_value == 0.99:
    print(check+"'s UnitPrice is "+ chk_value.__str__())


The Office's UnitPrice is 1.99
Aquaman's UnitPrice is 1.99
Pearl Jam's UnitPrice is 0.99
Lost's UnitPrice is 1.99


## SQL Aggregation

Di `pandas`, kita bisa menggunakan ketiga method berikut untuk membuat tabel agregasi:

- `.crosstab()`
- `.pivot_table()`
- `.groupby()`

Sedangkan di SQL, kita menggunakan statement `GROUP BY` yang diletakkan setelah `SELECT ... FROM ...`.

Misal kita ingin mengetahui top 5 `CustomerId` berdasarkan jumlah besaran transaksinya (`Total`) dan juga tampilkan banyak transaksinya:

In [54]:
top_cust = pd.read_sql_query("SELECT CustomerId, SUM(Total) AS TotalValue, \
                              COUNT(InvoiceId) AS Purchases \
                              FROM invoices \
                              GROUP BY CustomerId \
                              ORDER BY TotalValue DESC \
                              LIMIT 5", conn, index_col='CustomerId')
top_cust

TotalValue  Purchases
CustomerId                       
6                49.62          7
26               47.62          7
57               46.62          7
45               45.62          7
46               45.62          7

In [55]:
# re-create the table above


Perhatikan bagaimana query di atas dapat mengambil 5 customer teratas dari tabel `invoices`:

1. Data dikelompokkan berdasarkan `CustomerId` dengan statement `GROUP BY`
2. Terdapat dua fungsi agregasi yang digunakan: `SUM()` dan `COUNT()`, masing-masing mengagregasikan kolom `Total` dan `InvoiceId`. Berikut adalah fungsi agregasi yang sering digunakan: `SUM`,` AVG`, `COUNT`, `MIN`, dan `MAX`.
3. Statement `ORDER BY` ditambahkan untuk mengurutkan tabel berdasarkan kolom `TotalValue` secara `DESC`.

**Note**:

- SQL Statement `GROUP BY` ekuivalen dengan `.groupby()` pada `pandas`
- SQL Statement `ORDER BY` ekuivalen dengan `.sort_values()` pada `pandas`

### ❓ Knowledge Check

Seorang produser musik ingin menciptakan lagu dengan mempertimbangkan genre yang akan laku di market. Sebelum mulai, ia terlebih dahulu melakukan riset tentang genre musik apa yang sedang hits saat ini. Sebagai seorang data analyst, Anda diminta untuk membantu produser musik dalam membuat keputusan yang tepat.

**Pertanyaan:** Tampilkan top 5 genre musik yang memiliki jumlah penjualan terbanyak, sertakan juga dengan total quantity yang terjual.

<br>

<details>
    <summary><i>👉 Klik di sini untuk hint</i></summary>
    
Berikut adalah code yang dapat Anda gunakan untuk mengetahui `genres` paling populer dari semua invoice sales.

```python
top_genre = pd.read_sql_query(
    """
    SELECT
    genres.GenreId,
    genres.Name,
    ___(invoices.Total),
    ___(invoice_items.Quantity)
    FROM invoices
    LEFT JOIN ___ ON ___
    LEFT JOIN ___ ON ___
    LEFT JOIN ___ ON ___
    GROUP BY ___
    ORDER BY ___
    """,
    conn,
    index_col='GenreId'
```
</details>

In [70]:
cursor = conn.cursor() 
cursor.execute('''
    INSERT INTO tracks (Name,AlbumId,MediaTypeId,GenreId,Composer,Milliseconds,Bytes,UnitPrice)
    VALUES ("junkdata",1,1,1,"junkdata",1,1,5)
''')
conn.commit() 

In [69]:

# cursor.execute('''
#     DELETE FROM tracks WHERE Name='junkdata'
# ''')
# conn.commit() 

In [77]:
top_genre = pd.read_sql_query('''
          SELECT 
              *
          FROM tracks 
          WHERE Name='junkdata'
          ''', conn)
top_genre

TrackId      Name  AlbumId  MediaTypeId  GenreId  Composer  Milliseconds  \
0     3513  junkdata        1            1        1  junkdata             1   

   Bytes  UnitPrice  
0      1          5

In [78]:
top_genre = pd.read_sql_query('''
          SELECT 
              *
          FROM invoice_items 
          LEFT JOIN tracks ON tracks.trackid = invoice_items.trackid
          WHERE tracks.name='junkdata'
          ''', conn)
top_genre

Empty DataFrame
Columns: [InvoiceLineId, InvoiceId, TrackId, UnitPrice, Quantity, TrackId, Name, AlbumId, MediaTypeId, GenreId, Composer, Milliseconds, Bytes, UnitPrice, GenreId, Name]
Index: []

In [75]:
top_genre = pd.read_sql_query('''
          SELECT 
              genres.name as NameGenre
              , sum(invoice_items.UnitPrice*invoice_items.Quantity) as TotalValue
              , sum(invoice_items.Quantity) as TotalQty 
          FROM invoice_items 
          LEFT JOIN tracks ON tracks.trackid = invoice_items.trackid
          LEFT JOIN genres ON genres.genreid = tracks.genreid
          GROUP BY genres.name 
          ORDER BY TotalValue DESC 
          LIMIT 5''', conn)
top_genre

NameGenre  TotalValue  TotalQty
0                Rock      826.65       835
1               Latin      382.14       386
2               Metal      261.36       264
3  Alternative & Punk      241.56       244
4            TV Shows       93.53        47

In [49]:
top_genre = pd.read_sql_query('''
          SELECT 
              genres.name as NameGenre
              , invoice_items.UnitPrice*invoice_items.Quantity
              , invoice_items.Quantity
          FROM invoice_items 
          INNER JOIN tracks ON tracks.trackid = invoice_items.trackid
          INNER JOIN genres ON genres.genreid = tracks.genreid
      ''', conn)
top_genre

NameGenre  invoice_items.UnitPrice*invoice_items.Quantity  Quantity
0         Rock                                            0.99         1
1         Rock                                            0.99         1
2         Rock                                            0.99         1
3         Rock                                            0.99         1
4         Rock                                            0.99         1
...        ...                                             ...       ...
2235     Metal                                            0.99         1
2236     Metal                                            0.99         1
2237     Latin                                            0.99         1
2238     Latin                                            0.99         1
2239  TV Shows                                            1.99         1

[2240 rows x 3 columns]

In [40]:
# inv_items = pd.read_sql_query('''SELECT * from invoice_items''', conn)
# inv_items['Quantity'].unique()

array([1], dtype=int64)

💡 Panduan step-by-step:

1. Lakukan operasi `JOIN` tergantung dari kebutuhan kolom yang ingin dianalisis
2. Gunakan `GROUP BY` apabila ingin membuat tabel agregasi
3. Aplikasikan fungsi agregat di `SELECT`
4. Tambahkan kolom yang relevan untuk ditampilkan pada `SELECT`
5. Urutkan data menggunakan `ORDER BY`

## `WHERE` Statements

Sampai di sini, kita telah mempelajari beberapa statement SQL yang sering digunakan:

- `SELECT` statement
- SQL `JOIN`
- Aliasing
- SQL Aggregation dengan `GROUP BY`

Sekarang, kita akan melihat teknik untuk melakukan **conditional subsetting atau filter baris** menggunakan statement `WHERE` yang diikuti dengan **kondisi**.

### Logical Operator

- Kondisi `WHERE` dapat dikombinasikan dengan logical operator: `IS`, `AND`, `OR`, dan `NOT`:
    - `IS` sama saja seperti notasi matematis `=`
    - `IS NOT` seperti notasi `!=`
    
- Kondisi pada `WHERE` juga mendukung operator matematis seperti >, >=, <, dan <=

**Kasus**: kita ingin melakukan analisis terhadap semua data `invoices` yang terjadi di `BillingCountry` Germany, maka kita bisa menambahkan statement `WHERE` sebagai berikut:

In [ ]:
pd.read_sql_query(
    """
    SELECT *
    FROM invoices
    WHERE BillingCountry = 'Germany'
    """,
    conn)

Berlawanan dengan kasus di atas, misal kita ingin analisis semua data `invoices` dimana `BillingCountry` **selain** negara Germany:

**Kasus**: Misalkan kita hanya tertarik untuk menganalisis data yang terjadi di Amerika (yaitu `BillingCountry` Canada dan USA). Kira-kira bagaimana cara mengimplementasikannya dalam bentuk query SQL?

In [ ]:
# your code here


### `IN` Operator

Operator `IN` memungkinkan kita menentukan beberapa nilai untuk perbandingan. Misalnya, seperti pada kasus sebelumnya, kita ingin mengambil semua data `invoices` yang terjadi di negara (`BillingCountry`) Canada dan juga USA:

In [ ]:
north_america = pd.read_sql_query(
    """
    SELECT *
    FROM invoices
    WHERE BillingCountry IN ('Canada', 'USA')
    """,
    conn)

north_america.head()

### ❓ Knowledge Check


Ambillah seluruh data musik yang memiliki genre `Pop` dan `UnitPrice` dari tracknya adalah 0.99. Kemudian simpan ke object dataframe bernama `popmusic`.

**Pertanyaan:** Ada berapa baris pada dataframe `popmusic`?

<br>

<details>
    <summary><i>👉 Klik di sini untuk hint</i></summary>
    
```python
popmusic = pd.read_sql_query(
    """
    SELECT tracks.*, genres.Name AS GenreName
    FROM ___
    LEFT JOIN ___
    ON ___
    WHERE ___
    """,
    conn,
    index_col='TrackId'
)
```
</details>

In [73]:
# your code here
popmusic = pd.read_sql_query(
    """
    SELECT UnitPrice
    FROM tracks
    INNER JOIN genres ON genres.genreid = tracks.genreid
    WHERE genres.name='Pop' and UnitPrice = 0.99
    """,
    conn)

popmusic.size

48

### Operating Dates

Pada operasi kondisi pada statement `WHERE` sebelumnya, kita dapat mengambil semua data `invoices` pada negara Germany. Namun, juga umum untuk kita melakukan conditional query untuk mengambil data pada **rentang tanggal** tertentu.

Sebelum lanjut, mari kita lihat tipe data dari objek `germany` yang telah diperoleh sebelumnya:

In [69]:
germany = pd.read_sql_query("SELECT * FROM invoices WHERE BillingCountry = 'Germany'", conn)
germany.dtypes

InvoiceId              int64
CustomerId             int64
InvoiceDate           object
BillingAddress        object
BillingCity           object
BillingState          object
BillingCountry        object
BillingPostalCode     object
Total                float64
dtype: object

Perhatikan bahwa `InvoiceDate` dikenali sebagai tipe data `object`. **Method `pd.read_sql_query()` berperilaku seperti method `pd.read_csv()` dimana secara default membaca tipe data suatu kolom sebagai numerik dan objek.** Ini tidak berarti bahwa kolom tersebut disimpan menggunakan format string (umumnya dikenal sebagai `VARCHAR` dalam database SQL). Lihatlah skema tabel berikut:

In [70]:
invoices_table = pd.read_sql_query(
    """
    SELECT sql
    FROM sqlite_master
    WHERE name = 'invoices'
    """, conn)
print(invoices_table.loc[0,:].values[0])

CREATE TABLE "invoices"
(
    [InvoiceId] INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
    [CustomerId] INTEGER  NOT NULL,
    [InvoiceDate] DATETIME  NOT NULL,
    [BillingAddress] NVARCHAR(70),
    [BillingCity] NVARCHAR(40),
    [BillingState] NVARCHAR(40),
    [BillingCountry] NVARCHAR(40),
    [BillingPostalCode] NVARCHAR(10),
    [Total] NUMERIC(10,2)  NOT NULL,
    FOREIGN KEY ([CustomerId]) REFERENCES "customers" ([CustomerId]) 
		ON DELETE NO ACTION ON UPDATE NO ACTION
)


> Perlu diingat bahwa untuk DBMS yang berbeda-beda, maka cara mengambil skema tabel akan berbeda pula. Query di atas khusus untuk mengambil skema tabel dari database SQLite (`sqlite_master`).

Output dari query di atas termasuk dalam tipe **Data Definition Language (DDL)** yang digunakan untuk membuat tabel. Dengan membaca DDL, akan berguna untuk memahami skema tabel dari database sehingga kita dapat melakukan operasi yang sesuai. Pada skema tabel `invoices` terdapat informasi yang berguna seperti:

- `InvoiceId` sebagai primary key
- `InvoiceDate` disimpan sebagai tipe data `DATETIME` (format `YYYY-MM-DD HH:MM:SS`)
- `CustomerId` sebagai foreign key pada tabel `customers`

Jika Anda tidak disediakan dengan skema database dalam bentuk diagram, maka luangkan waktu untuk mempelajari setiap skema tabel melalui DDL-nya.

**Kasus**: Kita ingin meninjau penjualan (`invoices`) tahun lalu di seluruh negara pada tahun 2012 saja.

In [ ]:
invoice_2012 = pd.read_sql_query(
    """
    SELECT *
    FROM invoices
    WHERE ___
    """,
    conn,
    parse_dates='InvoiceDate'
)

### `BETWEEN` Operator

Melanjutkan kasus sebelumnya, apabila kita ingin menggunakan kondisi pada rentang tertentu, pendekatan yang umum digunakan adalah operator `BETWEEN`. Silahkan lengkapi code berikut dan lihat apakah data yang terambil sama seperti sebelumnya:

```python
invoice_2012 = pd.read_sql_query(
    """
    SELECT *
    FROM invoices
    WHERE ___ BETWEEN '___' AND '___'
    """,
    conn,
    parse_dates='InvoiceDate'
)

invoice_2012['InvoiceDate'].describe()
```

In [ ]:
# your code here


⚠️ **Hati-hati!**

Operator `BETWEEN` sejatinya adalah inclusive, di mana kondisi start dan endnya termasuk. Namun, saat kita membandingkan date (misal '2012-12-31') dengan datetime (pada kolom `InvoiceDate`), seakan-akan end tidak inclusive. Sebagai eksperimen, cobalah ganti kondisi pada statement `WHERE` menjadi:

```
WHERE InvoiceDate BETWEEN '2012-01-01' AND '2012-12-30'
```

Maka `InvoiceDate` pada tanggal `2012-12-30` tidak masuk.

Sebagai solusinya, lebih baik kita menambahkan komponen waktu (time) pada kondisi:
```
WHERE InvoiceDate BETWEEN '2012-01-01 00:00:00' AND '2012-12-30 00:00:00'
```

In [ ]:
# bukti bahwa operator BETWEEN inclusive pada start dan end
pd.read_sql_query(
    """
    SELECT * FROM invoices
    WHERE InvoiceId BETWEEN 2 AND 5
    """,
    conn
)

In [ ]:
# MASALAH: ketika hanya menggunakan '2012-12-30' maka data tanggal 30 Desember 2012 TIDAK masuk
pd.read_sql_query(
    """
    SELECT * FROM invoices
    WHERE InvoiceDate BETWEEN '2012-01-01' AND '2012-12-30'
    """,
    conn,
    parse_dates='InvoiceDate'
)['InvoiceDate'].describe()

In [ ]:
# SOLUSI: masukkan komponen TIME (00:00:00) agar data transaksi pada tanggal 30 Desember 2012 juga termasuk
pd.read_sql_query(
    """
    SELECT * FROM invoices
    WHERE InvoiceDate BETWEEN '2012-01-01 00:00:00' AND '2012-12-31 23:59:59'
    """,
    conn,
    parse_dates='InvoiceDate'
)['InvoiceDate'].describe()

> **Key takeaway:** Ketika ingin membuat sebuah kondisi `NAMA_KOLOM = NILAI`, pastikan `NILAI` memiliki tipe data yang sama seperti `NAMA_KOLOM`

### `LIKE` Operator

Operator `LIKE` sangat berguna jika kita perlu mencocokan bagian tertentu dari sebuah string daripada menggunakan operator sama dengan (`=`). `'107%'` yang Anda lihat dalam query ditujukan untuk mengekstrak nilai `BillingPostalCode` yang **dimulai** dengan angka 107. Ini sangat membantu ketika Anda ingin mengekstrak data hanya pada wilayah tertentu. Di negara Germany, kita akan tahu bahwa Wilmersdorf dan Tempelhof di Berlin memiliki kode pos dimulai dengan 107.

> **Note:** Karakter `%` disebut juga sebagai [wildcard character](https://www.w3schools.com/sql/sql_wildcards.asp)

In [ ]:
pd.read_sql_query(
    """
    SELECT * FROM invoices
    WHERE BillingCountry = 'Germany'
    AND BillingPostalCode LIKE '107%'
    """,
    conn
)

**Diskusi:**

- Jika `BillingPostalCode LIKE '107%'`, maka semua baris dengan kode pos **dimulai** angka 107 akan tampil.
- Jika `BillingPostalCode LIKE '%107'`, maka semua baris dengan kode pos **diakhiri** angka 107 akan tampil.

Menurut Anda apa yang akan muncul jika kita menggunakan `%` **sebelum dan sesudah** pola, yaitu `BillingPostalCode LIKE '%107%'`

#### ❓ Knowledge Check

**Kasus**: Pada 5 data pertama `customerinv`, kita bisa melihat kolom `Company` mungkin tidak dapat diandalkan karena bernilai `None`. Tetapi jika Anda memperhatikan kolom `Email`, Anda dapat melihat beberapa customer memiliki domain email `apple`, yang bisa menjadi indikator perusahaan mereka.

In [ ]:
customerinv = pd.read_sql_query(
    """
    SELECT FirstName, LastName, Email, Company,
    InvoiceId, InvoiceDate, BillingCountry, Total
    FROM invoices
    LEFT JOIN customers
    ON invoices.CustomerId = customers.CustomerId
    """,
    conn)

customerinv.head()

**Pertanyaan:** Berapa pelanggan yang bekerja di Apple Inc., jika kita menggunakan domain `Email` sebagai pengganti indikator `Company`?

- [ ] 412
- [ ] 49
- [ ] 7
- [ ] 14

Silahkan lengkapi kode berikut:

```python
applecust = pd.read_sql_query(
    """
    SELECT firstname, lastname, email, company,
    invoiceid, invoicedate, billingcountry, total
    FROM invoices
    LEFT JOIN customers
    ON invoices.___ = customers.___
    WHERE ___
    """,
    conn
)
```

In [ ]:
# your code here


## 📝 Summary

### ❓ Knowledge Check 1: SQL Statement Function

Kita telah mempelajari banyak statement pada SQL, silahkan mencocokan statement sesuai dengan kegunaannya:

**KEGUNAAN**

A. Memberikan nama lain pada tabel maupun kolom

B. Mengambil beberapa baris teratas dari tabel

C. Mengurutkan baris berdasarkan nilai pada kolom

D. Mengambil kolom dari sebuah tabel

E. Menggabungkan dua tabel menjadi satu tabel berdasarkan kolom penghubung

F. Filter baris

G. Membuat tabel agregasi

**STATEMENT**

1. `SELECT <nama_kolom> FROM <nama_tabel>`

2. `LIMIT <banyaknya_baris>`

3. `AS <nama_kolom_atau_baris>`

4. `<tabel_kiri> [LEFT|INNER] JOIN <tabel_kanan> ON <tabel>.key = <tabel>.key`

5. `GROUP BY <nama_kolom>`

6. `ORDER BY <nama_kolom> [ASC|DESC]`

7. `WHERE <kondisi>`

**JAWABAN**

1. 

2. 

3. 

4. 

5. 

6. 

7. 

### ❓ Knowledge Check 2: SQL Statement Structure

Susunlah 8 baris SQL statement berikut menjadi sebuah kerangka urutan syntax yang benar:

1. `GROUP BY <nama_kolom>`

2. `LIMIT <banyaknya_baris>`

3. `SELECT <nama_kolom> AS ...`

4. `[LEFT|INNER] JOIN <tabel_kanan> AS ...`

5. `ORDER BY <nama_kolom> [ASC|DESC]`

6. `FROM <nama_tabel> AS ...`

7. `WHERE <kondisi>`
    
8. `ON <tabel>.key = <tabel>.key`

> **JAWABAN**: Urutan yang benar adalah ...

### 💭 Dive Deeper

Toko musik digital ini ingin memberikan satu penghargaan ke `employees` yang telah bekerja keras melakukan penjualan di `BillingCountry` Amerika Selatan (yaitu negara **Argentina, Brazil, dan Chile**). Apabila penghargaan tersebut diberikan kepada `employees` berdasarkan **jumlah `Total`** penjualan, siapakah yang berhak mendapatkannya? Tampilkan **nama lengkap (`FirstName` dan `LastName`) beserta total penjualannya**.

In [ ]:
# your code here


# Under and Over Fetching

Konsep client-server pada database: Sedikit berbeda dengan `pandas` yang semua operasinya dilakukan di komputer lokal Anda. Ketika Anda bekerja dengan SQL, kemungkinan besar Anda memiliki database relasional yang disimpan terpusat pada server dan dapat diakses oleh beberapa client.

![](assets/clientserver.png)

Saat Anda melakukan query, sebenarnya Anda sedang menjalankan perintah untuk mengunduh data ke komputer lokal. Proses pengunduhan ini membutuhkan sumber daya dan perlu memanfaatkan alat secara efektif untuk meminimalkan biaya.

- **Over** fetching adalah kondisi dimana kita menarik data dari database **lebih dari** yang dibutuhkan, sehingga membutuhkan biaya dan waktu yang lebih.

- **Under** fetching adalah kondisi sebaliknya, dimana data yang ditarik **kurang dari** yang dibutuhkan, sehingga proses analisis tidak dapat dilakukan secara lengkap.

**Diskusi:**

Anda diminta untuk melakukan analisis terhadap semua penjualan (`invoices`) genre `Rock` pada tahun 2012. Pertimbangkan pertanyaan berikut:

- Apakah perlu bagi Anda untuk mengunduh semua tabel `tracks` ke komputer lokal?
- Apakah Anda melakukan filter baris terhadap `tracks` dengan genre `Rock` menggunakan SQL statement `WHERE` atau conditional subsetting `pandas`?
- Karena kita memerlukan informasi dari beberapa tabel, manakah cara yang lebih nyaman: melakukan query dengan `JOIN` atau melakukan `SELECT` secara terpisah dari database?

Cobalah membuat query yang paling optimum menurut Anda:

In [ ]:
# your code here


# (Optional) SQL Subquery

Dalam beberapa kasus tertentu, kita ingin melakukan filter baris berdasarkan syarat tertentu dimana nilai-nilai kondisi didapatkan dari hasil query lain. Apakah Anda masih ingat bagaimana kita mengambil semua pelanggan yang memiliki `invoices` dengan Total paling tinggi? (Bagian SQL Aggregation)

In [ ]:
# menampilkan top 5 customers dengan total pembelanjaan terbanyak
pd.read_sql_query(
    """
    SELECT CustomerId, SUM(Total) AS TotalValue,
    COUNT(InvoiceId) AS Purchases
    FROM invoices
    GROUP BY CustomerId
    ORDER BY TotalValue DESC
    LIMIT 5
    """,
    conn,
    index_col='CustomerId'
)

Misalnya dari tabel tersebut kita ingin menarik semua data `invoices` berdasarkan top 5 `customers`. Untuk melakukan hal tersebut, kita akan menggunakan subquery setelah statement `WHERE` menggunakan operator `IN`. Sebenarnya kita bisa saja menuliskan list `CustomerId` dalam kondisi secara **hard-code** seperti ini:

In [ ]:
# menampilkan invoices hanya untuk top 5 customers
top5cust_hardcode = pd.read_sql_query(
    """
    SELECT *
    FROM invoices
    WHERE CustomerId IN (6, 26, 57, 45, 46)
    """,
    conn)

top5cust_hardcode['CustomerId'].unique()

Namun cara hard-code seperti itu tidak disarankan. Bagaimana kalau database kita terus bertambah, sehingga bukan lagi `CustomerId` 6, 26, 57, 45, 46 yang merupakan top 5 customers? Maka dari itu, kita perlu menggunakan subquery agar **query kita lebih dinamis terhadap perubahan data**.

Untuk itu mari kita persiapkan **subquery** yang hanya mengembalikan list Top 5 `CustomerId` berdasarkan `Total` pembelian:

In [ ]:
# subquery
pd.read_sql_query(
    """
    SELECT CustomerId
    FROM invoices
    GROUP BY CustomerId
    ORDER BY SUM(Total) DESC
    LIMIT 5
    """,
    conn)

Subquery di atas kita gunakan untuk menggantikan query yang hard-code:

In [ ]:
top5cust_subquery = pd.read_sql_query(
    """
    SELECT *
    FROM invoices
    WHERE CustomerId IN (
        SELECT CustomerId
        FROM invoices
        GROUP BY CustomerId
        ORDER BY SUM(Total) DESC
        LIMIT 5
    )
    """,
    conn)

top5cust_subquery

Memang hasil `top5cust_hardcode` dengan `top5cust_subquery` akan sama persis, namun secara sintaks `top5cust_subquery` jauh lebih baik karena code lebih dinamis.

In [ ]:
top5cust_hardcode.shape

In [ ]:
# keuntungan: code lebih dinamis kalau ada perubahan data di masa depan
top5cust_subquery.shape